lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require 'erb'
+require 'json'
+require 'open3'
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    TRUE_VALUES = [true, 1, '1', 't', 'T', 'true', 'TRUE'].to_set
+    DEPLOY_METADATA_ANNOTATION = 'kubernetes-deploy.shopify.io/metadata'
+
+    def self.inherited(child_class)
+      DiscoverableResource.child_classes.add(child_class)
```

Comment:
Overriding allows us to distinguish dynamically created resources (discovery) versus statically defined resources (direct subclass of `KubernetesResource`).